api/context/rest.go

```diff
@@ -0,0 +1,247 @@
+		return nil, err
+	}
+
+	resp := listContextsResponse{}
+	if _, err := client.DoRequest(req, &resp); err != nil {
+		return nil, err
+	}
+	return &resp, nil
+}
+
+// Gets all pages of ListContexts
+func ListAllContextsWithRest(client *rest.Client, params ListContextsWithRestParams) ([]Context, error) {
+	contexts := []Context{}
+	for {
+		resp, err := ListContextsWithRest(client, params)
+		if err != nil {
+			return nil, err
+		}
+
+		contexts = append(contexts, resp.Items...)
+
+		if resp.NextPageToken == "" {
+			break
+		}
+
+		params.PageToken = resp.NextPageToken
+	}
+	return contexts, nil
+}
+
+type CreateContextWithRestParams struct {
+	Name  string `json:"name"`
+	Owner struct {
+		Type string `json:"type"`
+		Id   string `json:"id,omitempty"`
+		Slug string `json:"slug,omitempty"`
+	} `json:"owner"`
+}
+
+var orgOwnerType = "organization"
+
+// Creates a new context.
+// Binding to https://circleci.com/docs/api/v2/index.html#operation/createContext
+func CreateContextWithRest(client *rest.Client, params CreateContextWithRestParams) (*Context, error) {
+	if params.Owner.Id == "" && params.Owner.Slug == "" {
+		return nil, fmt.Errorf("to create a context, need either org ID or org slug, received none")
+	}
+	if params.Owner.Type == "" {
+		params.Owner.Type = orgOwnerType
+	}
+	if params.Owner.Type != orgOwnerType && params.Owner.Type != "account" {
+		return nil, fmt.Errorf("only owner.type values allowed to create a context are \"organization\" or \"account\", received: %s", params.Owner.Type)
+	}
+	if params.Owner.Id == "" && params.Owner.Slug != "" && params.Owner.Type == "account" {
+		return nil, fmt.Errorf("when creating a context, owner.type with value \"account\" is only allowed when using owner.id and not when using owner.slug")
+	}
+
+	u, err := client.BaseURL.Parse("context")
+	if err != nil {
+		return nil, err
+	}
+	req, err := client.NewRequest("POST", u, &params)
+	if err != nil {
+		return nil, err
+	}
+
+	var context Context
+	if _, err := client.DoRequest(req, &context); err != nil {
+		return nil, err
+	}
+
+	return &context, nil
+}
+
+type DeleteContextWithRestResponse struct {
+	Message string `json:"message"`
+}
+
+func DeleteContextWithRest(client *rest.Client, contextID string) (*DeleteContextWithRestResponse, error) {
+	u, err := client.BaseURL.Parse(fmt.Sprintf("context/%s", contextID))
+	if err != nil {
+		return nil, err
+	}
+	req, err := client.NewRequest("DELETE", u, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	resp := DeleteContextWithRestResponse{}
+	if _, err = client.DoRequest(req, &resp); err != nil {
+		return nil, err
+	}
+	return &resp, nil
+}
+
+type ListEnvVarsWithRestParams struct {
+	ContextID string
+	PageToken string
+}
+
+type listEnvVarsResponse struct {
+	Items         []EnvironmentVariable `json:"items"`
+	NextPageToken string                `json:"next_page_token"`
+}
+
+func ListEnvVarsWithRest(client *rest.Client, params ListEnvVarsWithRestParams) (*listEnvVarsResponse, error) {
```

Comment:
You could generalize for all return values. No need to return a pointer if you don't have anything hard to copy in the struct (like a mutex, a file or a network connection), or if you know the copying is a performance bottleneck. In many cases, not making it a pointer even lets the compiler do some cool optimizations.

Also, I think the return type should be exported, since it's returned from an exported func, and has exported fields.

```suggestion
func ListEnvVarsWithRest(client *rest.Client, params ListEnvVarsWithRestParams) (ListEnvVarsResponse, error) {
```

Reply:
This comment has sent me down a rabbit hole of when to:

* use pointers as receivers
* pass pointers as arguments
* return pointers from functions

Opinions vary, and many people inject different semantics into the simple `*` in Go. Optionals, mutability, pass-by-reference... Very different from Rust and its explicit `mut`, `&` references and `std::option`, or C++ and its many types of "smart pointers". We should probably take this offline and establish our own guide of "Pointer usage in Go". Hell, we could probably make a blog post out of it.

Do as you like for pointer types. You can resolve this, let's chat later.

Reply:
Nice, I don't know but maybe someone in CCI wrote some go practices for the company.